File: 백준/Gold/1027. 고층 건물/고층 건물.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer st;
    static StringBuilder sb = new StringBuilder();

    public static void main(String[] args) throws IOException {

        int N = Integer.parseInt(br.readLine());
        int[] building = new int[N + 1];

        st = new StringTokenizer(br.readLine());
        for (int i = 1; i <= N; i++) {
            building[i] = Integer.parseInt(st.nextToken());
        } // 입력완료

        int answer = 0;
        for (int i = 1; i <= N; i++) {
            int result = 0;
            double maxSlope = building[i]/0.5;
            for (int j = i - 1; j >= 1; j--) {
                if (maxSlope > ((double) building[i] - building[j]) / (i - j)) {
                    maxSlope = ((double) building[i] - building[j]) / (i - j);
                    result += 1;
                }
            }
            double minSlope = -building[i]/0.5;
            for (int j = i + 1; j <= N; j++) {
                if (minSlope < ((double) building[j] - building[i]) / (j - i)) {
                    minSlope = ((double) building[j] - building[i]) / (j - i);
                    result += 1;
                }
            }
            answer = Math.max(answer, result);
        } // end of for i
        
        System.out.println(answer);
    } // end of main

} // end of class

File: 백준/Gold/1027. 고층 건물/고층 건물.py
"""
완탐

각 건물마다 왼쪽과 오른쪽에 보이는 건물의 개수를 구한다
건물의 개수는 기울기로 구함
"""
import sys

input = lambda: sys.stdin.readline().strip()

n = int(input())
lst = list(map(int, input().split()))

answer = 0
for i in range(n):
    result = 0
    max_slope = lst[i] / 0.5
    for j in range(i)[::-1]:
        if max_slope > (lst[i] - lst[j]) / (i - j):
            max_slope = (lst[i] - lst[j]) / (i - j)
            result += 1
    min_slope = -lst[i] / 0.5
    for j in range(i + 1, n):
        if min_slope < (lst[j] - lst[i]) / (j - i):
            min_slope = (lst[j] - lst[i]) / (j - i)
            result += 1
    answer = max(answer, result)

print(answer)